cmd/snap-bootstrap/partition/encrypt.go

```diff
@@ -0,0 +1,144 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package partition
+
+import (
+	"bytes"
+	"crypto/rand"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+
+	"github.com/snapcore/snapd/osutil"
+)
+
+var (
+	tempFile = ioutil.TempFile
+)
+
+// The LUKS master key is 64 bytes long
+const (
+	masterKeySize = 64
+)
+
+// EncryptionKey holds the LUKS master key.
+type EncryptionKey [masterKeySize]byte
+
+func NewEncryptionKey() (EncryptionKey, error) {
+	var key EncryptionKey
+	_, err := rand.Read(key[:])
```

Comment:
Importing `crypto/rand` should provide us a cryptographically secure random number generator, but I'll check this with security and see what they recommend.

Reply:
I think we should at least put a comment explaining why be believe this is ok.

Reply:
@chrisccoulson, is this kind of approach ok security-wise or should we let cryptsetup generate its own master key and then extract it to a file (and re-use it in the next device if they're to share the same key)? I don't like the intermediate file in this case, but we already have a similar situation where a temporary file is required when opening the device.

Reply:
I think I said verbally that this approach looks ok, but I've taken a more thorough look just to make sure.

Go's rand.Reader on Linux is essentially just a wrapper that reads bytes directly from /dev/urandom.

For generating random numbers for keys, cryptsetup has the option of reading from either /dev/random or /dev/urandom depending on whether the --use-random or --use-urandom options are supplied at runtime. In the absence of either or those options, it uses the compiled in default which is currently /dev/urandom because our cryptsetup package isn't configured with --enable-dev-random.

So in this context, the approaches are equivalent.